plugin/advisor/mysql/advisor_table_pk_type.go

```diff
@@ -0,0 +1,237 @@
+package mysql
+
+// Framework code is generated by the generator.
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/parser/types"
+
+	"github.com/bytebase/bytebase/plugin/advisor"
+	"github.com/bytebase/bytebase/plugin/advisor/catalog"
+	"github.com/bytebase/bytebase/plugin/advisor/db"
+)
+
+var (
+	_ advisor.Advisor = (*TablePkTypeAdvisor)(nil)
+	_ ast.Visitor     = (*tablePkTypeChecker)(nil)
+)
+
+func init() {
+	advisor.Register(db.MySQL, advisor.MySQLTablePKType, &TablePkTypeAdvisor{})
+	advisor.Register(db.TiDB, advisor.MySQLTablePKType, &TablePkTypeAdvisor{})
+}
+
+// TablePkTypeAdvisor is the advisor checking for correct type of PK.
+type TablePkTypeAdvisor struct {
+}
+
+// Check checks for correct type of PK.
+func (*TablePkTypeAdvisor) Check(ctx advisor.Context, statement string) ([]advisor.Advice, error) {
+	stmtList, errAdvice := parseStatement(statement, ctx.Charset, ctx.Collation)
+	if errAdvice != nil {
+		return errAdvice, nil
+	}
+
+	level, err := advisor.NewStatusBySQLReviewRuleLevel(ctx.Rule.Level)
+	if err != nil {
+		return nil, err
+	}
+	checker := &tablePkTypeChecker{
+		level:            level,
+		title:            string(ctx.Rule.Type),
+		line:             make(map[string]int),
+		database:         ctx.Database,
+		tablesNewColumns: make(map[string]columnNameToColumnDef),
+	}
+
+	for _, stmt := range stmtList {
+		(stmt).Accept(checker)
+	}
+
+	if len(checker.adviceList) == 0 {
+		checker.adviceList = append(checker.adviceList, advisor.Advice{
+			Status:  advisor.Success,
+			Code:    advisor.Ok,
+			Title:   "OK",
+			Content: "",
+		})
+	}
+	return checker.adviceList, nil
+}
+
+type columnNameToColumnDef map[string]*ast.ColumnDef
+type tableNewColumn map[string]columnNameToColumnDef
+
+func (t tableNewColumn) set(tableName string, columnName string, colDef *ast.ColumnDef) {
+	if _, ok := t[tableName]; !ok {
+		t[tableName] = make(map[string]*ast.ColumnDef)
+	}
+	t[tableName][columnName] = colDef
+}
+
+func (t tableNewColumn) get(tableName string, columnName string) (colDef *ast.ColumnDef, ok bool) {
+	if _, ok := t[tableName]; !ok {
+		return nil, false
+	}
+	col, ok := t[tableName][columnName]
+	return col, ok
+}
+
+type tablePkTypeChecker struct {
+	adviceList       []advisor.Advice
+	level            advisor.Status
+	title            string
+	line             map[string]int
+	database         *catalog.Database
+	tablesNewColumns tableNewColumn
+}
+
+type pkData struct {
+	table      string
+	column     []string
+	columnType []string
+	line       int
+}
+
+// Enter implements the ast.Visitor interface.
+func (v *tablePkTypeChecker) Enter(in ast.Node) (ast.Node, bool) {
+	var pkDataList []pkData
+	switch node := in.(type) {
+	case *ast.CreateTableStmt:
+		tableName := node.Table.Name.String()
+		pds := v.addNewColumns(tableName, node.OriginTextPosition(), node.Cols)
```

Comment:
For columns and constraints in CREATE TABLE statement, we can get the line for them, so use `node.Cols[x].OriginTextPosition()` instead of `node.OriginTextPosition()`
Related code https://sourcegraph.com/github.com/bytebase/bytebase/-/blob/plugin/parser/utils.go?L57-60